### ISSUE TITLE
Different inference results for rectangular vs square inputs in YOLO26

### ISSUE THREAD

**Issue by SugarSol** [ultralytics/ultralytics#24428]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

**Description**
I found that inference results are inconsistent between rectangular inference and default square inference.
The following two inference calls produce different detection results on the same image:
`
from ultralytics import YOLO

model = YOLO("./yolo26n.pt")
results = model("../../coco128/train2017/000000000049.jpg")
results = model("../../coco128/train2017/000000000049.jpg", imgsz=(384, 672))
results[0].save()
`
**Log**
image 1/1 /home/easyits/ultralytics/tools/../../coco128/train2017/000000000049.jpg: 640x512 4 persons, 2 horses, 30.8ms
Speed: 18.4ms preprocess, 30.8ms inference, 11.7ms postprocess per image at shape (1, 3, 640, 512)
image 1/1 /home/easyits/ultralytics/tools/../../coco128/train2017/000000000049.jpg: 384x320 1 person, 1 horse, 1 potted plant, 26.6ms
Speed: 1.2ms preprocess, 26.6ms inference, 0.5ms postprocess per image at shape (1, 3, 384, 320)
**yolo check**
Ultralytics 8.4.24 🚀 Python-3.8.10 torch-1.9.1+cu111 CUDA:0 (Tesla V100-SXM2-32GB, 32494MiB)
Setup complete ✅ (36 CPUs, 157.1 GB RAM, 930.5/1007.9 GB disk)

OS                     Linux-4.18.0-348.7.1.el8_5.x86_64-x86_64-with-glibc2.17
Environment            Linux
Python                 3.8.10
Install                pip
RAM                    157.14 GB
Disk                   930.5/1007.9 GB
CPU                    Intel Xeon Gold 6133 CPU @ 2.50GHz
CPU count              36
GPU                    Tesla V100-SXM2-32GB, 32494MiB
GPU count              4
CUDA                   11.1

numpy                  ✅ 1.24.4>=1.23.0
matplotlib             ✅ 3.7.5>=3.3.0
opencv-python          ✅ 4.10.0.84>=4.6.0
pillow                 ✅ 10.4.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.10.1>=1.4.1
torch                  ✅ 1.9.1+cu111>=1.8.0
torchvision            ✅ 0.10.1+cu111>=0.9.0
psutil                 ✅ 6.0.0>=5.8.0
polars                 ✅ 1.8.2>=0.20.0
ultralytics-thop       ✅ 2.0.19>=2.0.18
torch                  ✅ 1.9.1+cu111!=2.4.0,>=1.8.0; sys_platform == "win32"

### Additional

I tested the same behavior using official YOLOv8 pretrained weights and did not observe this issue.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SugarSol, thank you for reporting this and for sharing the comparison details for YOLO26 inference 🚀 This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report related to different results between rectangular and square inference, please provide a complete [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if possible. Your snippet is helpful 👍, but an MRE should ideally include:
- The exact model file used
- The exact image used
- The full inference script
- Any custom arguments such as `conf`, `iou`, `rect`, `agnostic_nms`, or device settings
- The exact output images or saved predictions for both runs
- Confirmation that the behavior persists after upgrading

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

This is normal. You can get different prediction output if you're predicting on different `imgsz`. You should use the same `imgsz` as training.

**Comment by SugarSol** [comment]
Thanks for your reply.

I used the official pretrained `yolo26n` weights for inference, and the prediction differences between rectangular inference and default square inference are quite large.

I also performed the same comparison using official YOLOv8 pretrained weights under the same conditions, and the outputs were nearly identical.

Is this behavior still considered normal?

**Comment by Y-T-G** [comment]
Yes. The predictions you ran are at significantly different input sizes. The first prediction ran at `(1, 3, 640, 512)` while the second ran at `(1, 3, 384, 320)`. The first is closer to the original training `imgsz`, so it's more accurate.

**Comment by Y-T-G** [comment]
Also both inference are rectangular in your run. 

To use square inference, you need to use `rect=False`.

**Comment by SugarSol** [comment]
Thanks for the clarification.

I understand that different input resolutions can affect prediction accuracy. However, my main concern is that YOLO26 appears to be much more sensitive to rectangular inference compared to YOLOv8 under the same setup.

For example:

* YOLOv8 pretrained models produce nearly identical results between square and rectangular inference.
* `yolo26n` shows much larger differences under the same conditions.

Also, Ultralytics currently allows rectangular `predict` and `export` imgsz values, for example:

```bash
model.export(format="onnx", imgsz=(384, 672))
```

In my deployment scenario, the target hardware has padding limitations, so I cannot deploy with a 640x640 square input and must use rectangular shapes such as 672x384.

So I would like to confirm:

Is YOLO26 intentionally designed to work best only near the training square imgsz, or could this indicate that the model is more sensitive to aspect ratio changes than previous YOLO versions?

**Comment by glenn-jocher** [comment]
Thanks — not square-only, but your current test is still comparing `640x512` vs `384x320`: with default `rect=True`, a tuple `imgsz` uses minimal padding and does not force full `384x672` for single-image inference, as shown in the [predict docs](https://docs.ultralytics.com/modes/predict/); for a true fixed `672x384` deployment check, please run `predict()`/`val()` with `rect=False` and export the same `imgsz`, and if the gap remains share that `rect=False` MRE since YOLO26 is an end-to-end model and is not directly comparable to YOLOv8’s post-processing path ([end-to-end detection](https://docs.ultralytics.com/guides/end2end-detection/)).

**Comment by SugarSol** [comment]
Thanks for the explanation. I reran the tests using `rect=False` to enforce fixed input shapes and also compared `end2end=True/False`.

Here are the results using the official `yolo26n.pt` pretrained weights on the same image:

### 1. Fixed 640x640, end-to-end inference

```python
model = YOLO("./yolo26n.pt")
results = model(
    "../../coco128/train2017/000000000049.jpg",
    rect=False
)
```

Output:

```text
image 1/1 /home/easyits/ai/wpf/Employee/100/ultralytics/tools/../../coco128/train2017/000000000049.jpg: 640x640 4 persons, 2 horses, 10.4ms
Speed: 4.5ms preprocess, 10.4ms inference, 8.0ms postprocess per image at shape (1, 3, 640, 640)
```

---

### 2. Fixed 640x640, non-end-to-end inference

```python
model = YOLO("./yolo26n.pt")
results = model(
    "../../coco128/train2017/000000000049.jpg",
    end2end=False,
    rect=False
)
```

Output:

```text
image 1/1 /home/easyits/ai/wpf/Employee/100/ultralytics/tools/../../coco128/train2017/000000000049.jpg: 640x640 4 persons, 2 horses, 9.5ms
Speed: 4.0ms preprocess, 9.5ms inference, 9.2ms postprocess per image at shape (1, 3, 640, 640)
```

---

### 3. Fixed 384x672, end-to-end inference

```python
model = YOLO("./yolo26n.pt")
results = model(
    "../../coco128/train2017/000000000049.jpg",
    imgsz=(384, 672),
    rect=False
)
```

Output:

```text
image 1/1 /home/easyits/ai/wpf/Employee/100/ultralytics/tools/../../coco128/train2017/000000000049.jpg: 384x672 1 person, 2 horses, 1 potted plant, 10.1ms
Speed: 2.8ms preprocess, 10.1ms inference, 7.9ms postprocess per image at shape (1, 3, 384, 672)
```

---

### 4. Fixed 384x672, non-end-to-end inference

```python
model = YOLO("./yolo26n.pt")
results = model(
    "../../coco128/train2017/000000000049.jpg",
    imgsz=(384, 672),
    end2end=False,
    rect=False
)
```

Output:

```text
image 1/1 /home/easyits/ai/wpf/Employee/100/ultralytics/tools/../../coco128/train2017/000000000049.jpg: 384x672 1 person, 3 horses, 1 potted plant, 10.7ms
Speed: 2.2ms preprocess, 10.7ms inference, 9.5ms postprocess per image at shape (1, 3, 384, 672)
```

The prediction differences between 640x640 and 384x672 still appear quite large even with fixed-shape inference and with both end-to-end and non-end-to-end modes.

Since many deployment targets require rectangular inputs, I would like to confirm whether this level of aspect-ratio sensitivity is expected behavior for YOLO26 models, and whether training with a matching rectangular aspect ratio is recommended for deployment scenarios such as 384x672.

The test image is from the official COCO dataset (coco128/train2017/000000000049.jpg).
The tests were performed using the official yolo26n.pt pretrained weights.

Also, during training/export, Ultralytics currently reports:
```bash
WARNING ⚠️ updating to 'imgsz=672'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer
```
This suggests that rectangular inference/export is an officially supported workflow, which is why I wanted to better understand whether the observed sensitivity to rectangular inputs is expected for YOLO26 models.

**Comment by Y-T-G** [comment]
> Is YOLO26 intentionally designed to work best only near the training square imgsz, or could this indicate that the model is more sensitive to aspect ratio changes than previous YOLO versions?

All the YOLO models work best closer to the `imgsz` they are trained at. From what I see, you're testing on a single image which isn't quite enough to draw a conclusions. For proper conclusion, I would recommend that you test on 100s of images and compare the differences. A single image can just be a fluke.

Aspect ratio is always preserved in YOLO prediction. All the predictions use same aspect ratio. Only the padding would be different.
